### PR TITLE
Flatten label sections

### PR DESCRIPTION
**Pull request by TerryChau**
Changes:

1. This pull request will load sections in a flat manner. See [flat_label_section.png](https://github.com/pharmaDB/dailymed_data_processor/blob/4c355691a3ae2663faedd54db7ece4428573570f/assets/flat_label_section.png)
This manner is preferred for work downstream, since every depth creates more complexity when attempting to batch text for NLP. I need to be able to grab all text, do nlp on all text, do similarity, then associate similarity back to all text. It's way more complex if sections could have subsections.

2. LABEL_SECTIONS now only needs to include the section titles that were highlighted in yellow here [Sections By T. Marman](https://docs.google.com/document/d/1RgGzXZTaL4_2jNy_ZrIMyQwDJJT0kxFk/edit). All subsections will be pull. The sections and subsections titles will include numbers. If we want, I can add an extra key:value pair such as "subtitle":true to each sub-document in json to indicate if title is subtitle.

This is also desirable, since the names of the subsections changes, just as the title of the section. This way we only have to track all variants of the title.

3. unicodedata.normalize() was included to remove nbsp in xml that was preventing some labels from matching. Now all labels in resource/set_id.json gets filled.

4. Included a parent:value pair.

==
Please fix .env if you like and set back to mock_data.

### CODE REVIEW

File: db/mongo.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 
 
 class MongoClient:
+    db_client=None
```
Comment (krithika369) [inline db/mongo.py]: Why is this needed?

File: Pipfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[[source]]
```
Comment (krithika369) [inline Pipfile]: There is already a `requirements.txt` in the repo. Perhaps we can remove that if the Pipfile is added, and update the running instructions in the README.

File: spl/labels.py [hunk 3]
```diff
@@ -186,55 +198,90 @@ def __get_application_numbers(self, set_id, bs_content):
...
+            )
+
+            # test if any string matches is in LABEL_SECTIONS_EXACT
+            title_string_in_LABEL_SECTION_EXACT = any(
+                label_section.lower() == get_xml_title(title.text).lower()
+                for label_section in SplHistoricalLabels.LABEL_SECTIONS_EXACT
+            )
+
+            # test if any substring is in LABEL_SECTION_BLACKLIST
+            title_substring_in_LABEL_SECTION_BLACKLIST = any(
```
Comment (krithika369) [inline spl/labels.py]: Why is `LABEL_SECTIONS_BLACKLIST` required? Is it not enough to check the requires labels list?

File: spl/labels.py [hunk 4]
```diff
@@ -186,55 +198,90 @@ def __get_application_numbers(self, set_id, bs_content):
...
+                    )
+                    labels.append(
+                        {
+                            "name": parent,
+                            "text": sibling_text,
+                            "parent": None,
+                        }
+                    )
+                    # loop through all subtitles and add subtitle and text
+                    for j in range(1, len(subtitles)):
```
Comment (krithika369) [inline spl/labels.py]: For subtitle in subtitles ... ?

File: spl/labels.py [hunk 5]
```diff
@@ -186,55 +198,90 @@ def __get_application_numbers(self, set_id, bs_content):
...
-        def get_xml_text(navigable_title):
-            # Process the label text
-            label_text = (
-                re.sub(r"\n{3,}", "\n\n", navigable_title.get_text())
-                # Remove initial / trailing white space
-                .lstrip().rstrip()
-            )
-            return label_text
+        def get_xml_text(text):
+            # Process the label text; normalize remove nbsp
```
Comment (krithika369) [inline spl/labels.py:193]: The input to this function is `.text` from the navigable string. Does it handle nested strings? Perhaps under a different tag like `table` ?

That's why the `get_text()` method was being used here.

File: spl/labels.py [hunk 6]
```diff
@@ -186,55 +198,90 @@ def __get_application_numbers(self, set_id, bs_content):
...
-                                                    sub_title.parent
-                                                ),
-                                            }
-                                        )
-                                        # Remove the sub title from the main title
-                                        sub_title.parent.decompose()
-
-                        # Add the section text, along with the sub-sections
+        i = 0
+        while i < len(titles):
```
Comment (krithika369) [inline spl/labels.py]: For title in titles?

Comment (krithika369) [inline spl/labels.py]: This section feels a little verbose, I think it can be refactored to extract some common logic into a function (eg: there is a single `labels.append()`)

**Review by krithika369 — CHANGES_REQUESTED** [review]
@TerryChau thanks for the changes and sorry for the delay on the review. I added some comments to the PR. Let me know if you have any concerns with any of these and if not, I can make a commit to address them given that you're busy with the ML.

File: db/mongo.py (hunk 1 above)
Comment (TerryChau) [reply db/mongo.py]: Can be removed

File: Pipfile (hunk 2 above)
Comment (TerryChau) [reply Pipfile]: Go ahead and remove the Pipfile.

File: spl/labels.py (hunk 3 above)
Comment (TerryChau) [reply spl/labels.py]: Maybe there might be a better way to do this without blacklists, but `Indications` from LABEL_SECTIONS would naturally cause `Contraindications`, which should be excluded from the sections of interest as identified by Dr. Marman, to be included with the sections collected.  I want to keep the number with the section title.  So if the section title is '2 Indications', I want to keep the entirety of the section title.  The number is useful for diffing as it can be used to track where new sections are added or removed relative to other sections.  The best way that I came up to do this is to test if the substring 'indications" is in the title.  However 'indications' is also in 'contraindications'.   To exclude contraindications,  a blacklist is used.

File: spl/labels.py (hunk 4 above)
Comment (TerryChau) [reply spl/labels.py]: I would keep this as is.  Subtitles technically includes the parent title, because of `subtitles = title.parent.find_all("title")`.  So the first subtitle is the parent title.  Hence looping from the second subtitle to the last one.  I believe this is a quirk with how the xml is arranged.  The best way to grab subtitles is to go to parent and get all titles within parent, and exclude the first one because it is the title of the parent.  An alternative to grab titles is to do child sibling child recursions, which is more complex.

File: spl/labels.py (hunk 5 above)
Comment (TerryChau) [reply spl/labels.py:193]: I believe these functions are all the same.  See https://stackoverflow.com/questions/52093875/gettext-vs-text-vs-get-text.  As far I as know text is a property that calls get_text()

File: spl/labels.py (hunk 6 above)
Comment (TerryChau) [reply spl/labels.py]: I understand pythonic loops are cleaner, but in this case I am not doing it with pythonic loops since I didn't want to recursively find subtitles.  Finding sibling and child titles can be hard since some subtitles are nested in multiple sibling and multiple childs tags with no uniform tags to identify children.  It is much easier to grab all titles within the parent of the title and recognized that anything that is not the first title is a subtitle. 
Let's say t are titles and s are subtitles, and we have the following structure.
`t1 [t2 s1 s2 s3] t3`
To grab subtitle s1 s2 s3, it is best to just get parent of t2 and get all titles within that parent.  The first title is a title and the remaining titles are subtitles.
It is possible to traverse all titles from t1 t2 s1 s2 s3 to t3 in order, but that would require that for each title, parents are traverse to test if the title is subtitle or an actual title, which may be complex endeavor, since the parent title is not always directly above the child but may be multiple level above child.  To do the test, every subtitle might have to traverse all the way up to the first title.

Comment (TerryChau) [reply spl/labels.py]: See comment above.  However feel free to refactor.

File: spl/labels.py (hunk 5 above)
Comment (krithika369) [reply spl/labels.py:193]: I see.. Thanks for the clarification.

File: spl/labels.py (hunk 6 above)
Comment (krithika369) [reply spl/labels.py]: Thanks for the explanation.

```
 since I didn't want to recursively find subtitles
```
In both cases (the old and new code), the find is not 'recursive' per se. It is as you have described here (grab all titles under the section title matched). I am not questioning the matching, just the use of indices for access which makes the code look clunky. If excluding the first match, it could be:

```
for subtitle in subtitles[1:]
```

I'll update it accordingly.

File: spl/labels.py (hunk 3 above)
Comment (krithika369) [reply spl/labels.py]: @TerryChau I've refactored the code a little bit. And I've verified that the result is the same with your version of the code and my latest commit, for the example set ids you had quoted.
```
    "762b51be-1893-4cd1-9511-e645fc420d3a",
    "1b5e2860-6855-4a65-8bbc-e064172a1adf",
    "b5cee013-000f-4e35-a284-1f58add31b4d"
```

The only difference is with Set ID `762b51be-1893-4cd1-9511-e645fc420d3a`. The sub-section `6.1 Clinical Trials Experience` was being matched because of the string contains comparison. I've changed this up to use an 'exact match', after cleaning the text:

```
# For matching:
            # * Convert title text to upper case
            # * Replace & with AND, tab with space
            # * Clean up Unicode characters, digits, punctuation
            title_match_text = cleantext.clean(
                title.text.upper().replace("&", "AND").replace("\t", " "),
                lower=False,
                no_line_breaks=True,
                no_punct=True,
                no_numbers=True,
                no_digits=True,
                replace_with_number="",
                replace_with_digit="",
            )
```

**Comment by krithika369** [comment]
@TerryChau Thanks for these changes. I'll merge them in. 🙏🏽

File: spl/labels.py (hunk 4 above)
Comment (krithika369) [reply spl/labels.py]: Just made a simple change here: `for subtitle in subtitles[1:]`
